components/feature/search/src/test/java/mozilla/components/feature/search/middleware/SearchMiddlewareTest.kt

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.search.middleware
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import mozilla.components.browser.state.action.SearchAction
+import mozilla.components.browser.state.search.RegionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.support.test.ext.joinBlocking
+import mozilla.components.support.test.libstate.ext.waitUntilIdle
+import mozilla.components.support.test.robolectric.testContext
+import org.junit.After
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.util.Locale
+
+@RunWith(AndroidJUnit4::class)
+class SearchMiddlewareTest {
+    private lateinit var dispatcher: TestCoroutineDispatcher
+    private lateinit var originalLocale: Locale
+
+    @Before
+    fun setUp() {
+        dispatcher = TestCoroutineDispatcher()
+        originalLocale = Locale.getDefault()
+    }
+
+    @After
+    fun tearDown() {
+        dispatcher.cleanupTestCoroutines()
+
+        if (Locale.getDefault() != originalLocale) {
+            Locale.setDefault(originalLocale)
+        }
+    }
+
+    @Test
+    fun `Loads search engines for region`() {
+        val store = BrowserStore(
+            middleware = listOf(SearchMiddleware(testContext, ioDispatcher = dispatcher))
+        )
+
+        assertTrue(store.state.search.regionSearchEngines.isEmpty())
+
+        store.dispatch(SearchAction.SetRegionAction(
+            RegionState("US", "US")
+        )).joinBlocking()
+
+        dispatcher.advanceUntilIdle()
+        // Why?
+        Thread.sleep(1000)
+
+        store.waitUntilIdle()
```

Comment:
@csadilek I need a second pair of eyes here. I do not understand why this is not working without the sleep: I first dispatch the action blocking, then I wait for the dispatcher to be idle (which executes a task that loads the engines and dispatches a new action), then wait for the store to be idle (to wait for the search engines to be added). We have a bunch of other tests doing this. But here it fails without the sleep. With some logging I see the loading of the search engines not being completed yet and `advanceUntilIdle()` apparently not waiting for it.  It feels like I am missing something obvious?!

Reply:
👀

Reply:
@pocmo OK, I found the problem now. There's another nested dispatcher which is independent of the tests in `loadSearchEnginesFromList` (it used and independent IO dispatcher which we're not blocking on in the test). One min pushing up a fix.

Reply:
@pocmo fix to use the same dispatcher: https://github.com/csadilek/android-components/commit/4d41373082137765f8e191136f0453636e44d1f2

Reply:
Thank you! I still don't fully understand what's going on, I think. :)

I cherry-picked your patch and did fix the tests. The other tests that I wrote in the meantime still fail. Maybe there's a similar problem with the custom engine storage too. 🤔